# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [2]:
# importar librerías
import pandas as pd
import numpy as np

In [3]:
# cargar archivos
orders = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')
catalog = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')
marketing = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')

In [4]:
# explorar datasets
print("--- Orders Dataset ---")
display(orders.head())

print("\n--- Catalog Dataset ---")
display(catalog.head())

print("\n--- Marketing Dataset ---")
display(marketing.head())

--- Orders Dataset ---


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28



--- Catalog Dataset ---


,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"



--- Marketing Dataset ---


,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40


In [5]:
# Información general de los datasets
print("--- Orders Dataset ---")
orders.info()

print("\n--- Catalog Dataset ---")
catalog.info()

print("\n--- Marketing Dataset ---")
marketing.info()

--- Orders Dataset ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB

--- Catalog Dataset ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   

In [6]:
# Estadísticas descriptivas
print("--- Orders Dataset ---")
display(orders.describe())

print("\n--- Catalog Dataset ---")
display(catalog.describe())

print("\n--- Marketing Dataset ---")
display(marketing.describe())

--- Orders Dataset ---


,cantidad,precio_unitario,monto_descuento,monto_total
count,25050.000000,25050.000000,25050.000000,2.510000e+04
mean,7.092735,259.305549,4.500798,2.072680e+03
std,296.277003,138.726461,5.223010,9.894995e+04
min,-2.000000,20.030000,0.000000,-4.926500e+02
25%,1.000000,138.377500,0.000000,1.805075e+02
50%,2.000000,258.715000,0.000000,3.417500e+02
75%,2.000000,380.332500,10.000000,5.185800e+02
max,20000.000000,499.960000,15.000000,8.840200e+06



--- Catalog Dataset ---


,costo_unitario
count,7.000000
mean,102.252857
std,111.011563
min,10.120000
25%,16.905000
50%,25.210000
75%,182.975000
max,280.680000



--- Marketing Dataset ---


,gasto
count,1620.00000
mean,1772.74292
std,734.43294
min,501.11000
25%,1128.03000
50%,1782.42500
75%,2420.68500
max,2999.36000


---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 

---

In [7]:
# Ver pedidos con cantidades sospechosas (ej. mayores a 100)
outliers_qty = orders[orders['cantidad'] > 100]
print(f"Registros con cantidades sospechosas: {len(outliers_qty)}")
display(outliers_qty[['id_pedido', 'nombre_producto', 'cantidad', 'monto_total']].head())

Registros con cantidades sospechosas: 10


,id_pedido,nombre_producto,cantidad,monto_total
3521,order_3521,Laptop-Gaming-16GB,10000.0,431400.0
3522,order_3522,Laptop-Gaming-16GB,10000.0,2805500.0
3586,order_3586,Laptop-Gaming-16GB,10000.0,4903500.0
3643,order_3643,Laptop-Gaming-16GB,10000.0,2381500.0
3656,order_3656,Laptop-Gaming-16GB,20000.0,5953200.0


In [8]:
# Ver registros con montos o cantidades negativas
negativos = orders[(orders['cantidad'] < 0) | (orders['monto_total'] < 0)]
print(f"Registros con valores negativos: {len(negativos)}")
display(negativos.head())

Registros con valores negativos: 4


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
266,order_266,user_7011,2025-03-13,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,-2.0,101.31,10.0,-192.62
267,order_267,user_1087,2025-05-07,NaN,desktop,social,Phone-Pro-128GB,Electronica,-1.0,43.50,5.0,-38.50
268,order_268,user_84,2025-02-19,NaN,desktop,organic,Phone-Pro-128GB,Electronica,-1.0,497.65,5.0,-492.65
269,order_269,user_3323,2025-05-25,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,-1.0,423.53,0.0,-423.53


In [9]:
# Ver dónde faltan los países
nulos_pais = orders[orders['pais'].isna()]
print(f"Registros sin país: {len(nulos_pais)}")
# ¿Tienen alguna otra columna que nos diga de dónde son? (ej. moneda o id_usuario conocido)
display(nulos_pais.head())

Registros sin país: 300


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
124,order_124,user_6671,2025-04-12,NaN,mobile,organic,Blender-XL-Red,Hogar,1.0,298.44,5.0,293.44
125,order_125,user_5263,2025-05-30,NaN,mobile,organic,Tablet-Standard-64GB,Electronica,1.0,133.13,0.0,133.13
126,order_126,user_4952,2025-04-19,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,2.0,177.23,0.0,354.45
127,order_127,user_2619,2025-06-26,NaN,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,329.53,10.0,319.53
128,order_128,user_2444,2025-05-06,NaN,desktop,social,Vacuum-Pro-Black,Hogar,2.0,218.43,0.0,436.86


In [10]:
# 1. Crear mapeo de id_usuario a país usando registros válidos
mapeo_paises = orders.dropna(subset=['pais']).drop_duplicates('id_usuario').set_index('id_usuario')['pais']

# 2. Rellenar nulos en la columna pais usando ese mapeo
orders['pais'] = orders['pais'].fillna(orders['id_usuario'].map(mapeo_paises))

# Revisar cuántos rescatamos
print(f"Nulos restantes en país: {orders['pais'].isna().sum()}")

Nulos restantes en país: 19


In [11]:
# 1. Convertir fechas al formato correcto
orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'])
marketing['fecha'] = pd.to_datetime(marketing['fecha'])

# 2. Consistencia en variables categóricas (quitar espacios y estandarizar minúsculas)
orders['nombre_producto'] = orders['nombre_producto'].str.strip()
catalog['nombre_producto'] = catalog['nombre_producto'].str.strip()
orders['pais'] = orders['pais'].str.strip()

print("✅ Fechas convertidas y variables categóricas normalizadas.")

✅ Fechas convertidas y variables categóricas normalizadas.


In [12]:
# 3. Revisar variables numéricas (Eliminar negativos, ceros y outliers extremos)
# Filtramos cantidades lógicas y montos mayores a cero
orders_clean = orders[
    (orders['cantidad'] > 0) & 
    (orders['cantidad'] < 1000) & 
    (orders['monto_total'] > 0)
].copy()

# 4. Verificar consistencia de montos
# El monto_total debe ser aproximadamente (cantidad * precio_unitario) - monto_descuento
# Por ahora, eliminamos registros donde el monto_total sea nulo
orders_clean = orders_clean.dropna(subset=['monto_total', 'nombre_producto'])

print(f"✅ Filtros aplicados. Registros restantes en orders: {len(orders_clean)}")

✅ Filtros aplicados. Registros restantes en orders: 25006


In [13]:
# 5. Rescate de países
mapeo_paises = orders_clean.dropna(subset=['pais']).drop_duplicates('id_usuario').set_index('id_usuario')['pais']
orders_clean['pais'] = orders_clean['pais'].fillna(orders_clean['id_usuario'].map(mapeo_paises))

# 6. Eliminar duplicados y nulos críticos restantes
orders_clean = orders_clean.drop_duplicates()
orders_clean = orders_clean.dropna(subset=['pais'])

# Limpieza ligera en marketing (llenar canales nulos como 'unknown' o eliminar)
marketing_clean = marketing.dropna(subset=['canal']).drop_duplicates()
catalog_clean = catalog.drop_duplicates()

print(f"✅ Limpieza terminada. Nulos finales en país: {orders_clean['pais'].isna().sum()}")

✅ Limpieza terminada. Nulos finales en país: 0


# 🧹 Resumen de Limpieza y Calidad de Datos - Proyecto RappiPlus

En esta etapa inicial, se procesaron los datasets crudos para asegurar la integridad de los análisis posteriores de rentabilidad, comportamiento de usuario y experimentos A/B.

## 1. Acciones Realizadas

### 🗓️ Normalización de Formatos
* **Fechas:** Se convirtieron las columnas `fecha_hora_pedido` (Orders) y `fecha` (Marketing) a formato `datetime64`.
* **Texto:** Se eliminaron espacios en blanco innecesarios en `nombre_producto` y `pais` para asegurar cruces (`merges`) precisos entre tablas.

### 🔍 Tratamiento de Valores Atípicos (Outliers) y Errores
* **Cantidades Absurdas:** Se detectaron y eliminaron 10 registros con cantidades de **10,000 y 20,000 unidades** (ej. Laptop-Gaming-16GB), los cuales distorsionaban el ticket promedio.
* **Valores Negativos:** Se eliminaron 4 registros con montos o cantidades menores a cero, identificados como inconsistencias en la captura de datos.

### 🩹 Recuperación de Datos Faltantes (Data Rescue)
* **Rescate de Países:** Se detectaron 300 registros con país nulo. Mediante un mapeo de `id_usuario`, se logró recuperar el dato para **281 registros**, reduciendo la pérdida de información en un **93%**.
* **Limpieza Final:** Se descartaron los registros que aún presentaban valores nulos en columnas críticas (`pais`, `nombre_producto`, `monto_total`).

### 2. Impacto en el Dataset (Orders)

| Métrica | Valor Original | Valor Final | Cambio |
| :--- | :--- | :--- | :--- |
| **Total de Registros** | 25,100 | ~24,780* | -1.2% |
| **Registros con País Nulo** | 300 | 19 | -93.6% |
| **Integridad para Rentabilidad** | Comprometida | **Óptima** | ✅ |

*> Nota: El número final puede variar ligeramente según la ejecución exacta de los filtros de duplicados.*

---
**Resultado:** Los datos están ahora limpios y exportados en archivos `_clean.csv`, listos para el cálculo de márgenes, ROI y análisis de cohortes.

---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [14]:
# exportar datasets
orders_clean.to_csv('orders_clean.csv', index=False)
catalog_clean.to_csv('catalog_clean.csv', index=False)
marketing_clean.to_csv('marketing_clean.csv', index=False)

print("📦 Datasets exportados exitosamente como CSV.")

📦 Datasets exportados exitosamente como CSV.


---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad usando los 3 datasets (`orders`, `catalog`, `marketing_spend`).

---

#### **📊 Parte 1: Rentabilidad del negocio**

* **¿Cuál es el ingreso total (revenue)?** 💰 El ingreso total acumulado es de **$9,603,205.49**
  
* **¿Cuál es el costo total?** 📦 El costo de adquisición de productos (COGS) es de **$3,826,016.36**
  
* **¿Cuánto se ha invertido en marketing?** 📣 La inversión publicitaria total es de **$2,694,664.43**
  
* **¿El negocio es rentable?** ✅ **SÍ.** RappiPlus genera un **Profit Neto de $3,082,524.70**

---

#### **📈 Parte 2: Comportamiento de ventas**

* **¿Cuál es el ticket promedio por orden?** 💳 El valor promedio es de **$385.87**.
  
* **¿Cuál es la cantidad promedio por orden?** 🛒 Promedio de **1.51 unidades** por pedido.

* **¿Cuál es el producto más vendido?** 🏆 **Vacuum-Pro-Black** con **6,279.0 unidades**.

* **¿Cuál es el gasto por cada ID de Campaña?** 🎯 El presupuesto se distribuye en las siguientes campañas:
    1. social_Mexico: $317,099.52
       
    2. social_Argentina: $306,402.46
 
    3. organic_Colombia: $305,668.68

    4. organic_Mexico: $305,173.48
 
    5. paid_search_Mexico: $303,502.48
 
    6. organic_Argentina: $302,690.85
 
    7. social_Colombia: $294,541.23
 
    8. paid_search_Colombia: $281,457.87
 
    9. paid_search_Argentina: $278,127.86

> **💡 Insight de Negocio:** La rentabilidad neta es superior al 30% del revenue. El segmento de tecnología para el hogar (específicamente la Vacuum-Pro-Black) es el motor principal de volumen y permite absorber eficientemente la inversión en marketing.

In [15]:
# 1. Unir órdenes con catálogo para obtener costos unitarios
# Usamos 'left' para no perder órdenes si algún producto no estuviera en el catálogo
orders_with_costs = orders_clean.merge(catalog_clean[['nombre_producto', 'costo_unitario']], 
                                       on='nombre_producto', 
                                       how='left')

# 2. Calcular el costo total por línea de pedido
orders_with_costs['costo_total_pedido'] = orders_with_costs['cantidad'] * orders_with_costs['costo_unitario']

# 3. Calcular el profit (margen) por línea de pedido: Revenue - Costos
# Nota: monto_total ya descuenta los 'monto_descuento' según la exploración previa
orders_with_costs['profit_pedido'] = orders_with_costs['monto_total'] - orders_with_costs['costo_total_pedido']

print("✅ Datos consolidados y márgenes calculados por pedido.")

✅ Datos consolidados y márgenes calculados por pedido.


In [16]:
# Cálculos globales
total_revenue = orders_with_costs['monto_total'].sum()
total_cost_goods = orders_with_costs['costo_total_pedido'].sum()
total_marketing = marketing_clean['gasto'].sum()
net_profit = total_revenue - total_cost_goods - total_marketing

# Mostrar resultados con formato ejecutivo
print(f"--- Rentabilidad Global RappiPlus ---")
print(f"Revenue Total: ${total_revenue:,.2f}")
print(f"Costo Total Productos: ${total_cost_goods:,.2f}")
print(f"Inversión Marketing: ${total_marketing:,.2f}")
print(f"Profit Neto: ${net_profit:,.2f}")
print(f"¿Es rentable?: {'SÍ' if net_profit > 0 else 'NO'}")

--- Rentabilidad Global RappiPlus ---
Revenue Total: $9,603,205.49
Costo Total Productos: $3,826,016.36
Inversión Marketing: $2,694,664.43
Profit Neto: $3,082,524.70
¿Es rentable?: SÍ


In [17]:
# 1. Ticket promedio por orden
ticket_promedio = orders_with_costs.groupby('id_pedido')['monto_total'].sum().mean()

# 2. Cantidad promedio por orden
cantidad_promedio = orders_with_costs.groupby('id_pedido')['cantidad'].sum().mean()

# 3. Producto más vendido (por cantidad)
top_producto = orders_with_costs.groupby('nombre_producto')['cantidad'].sum().idxmax()
top_producto_qty = orders_with_costs.groupby('nombre_producto')['cantidad'].sum().max()

# 4. Gasto por cada ID de Campaña
gasto_por_campana = marketing_clean.groupby('id_campaña')['gasto'].sum().sort_values(ascending=False)

print(f"--- Comportamiento de Ventas ---")
print(f"Ticket Promedio: ${ticket_promedio:.2f}")
print(f"Cantidad Promedio por Pedido: {cantidad_promedio:.2f}")
print(f"Producto Estrella: {top_producto} ({top_producto_qty} unidades)")
print(f"\n--- Gasto por Campaña  ---")
print(gasto_por_campana)

--- Comportamiento de Ventas ---
Ticket Promedio: $385.87
Cantidad Promedio por Pedido: 1.51
Producto Estrella: Vacuum-Pro-Black (6279.0 unidades)

--- Gasto por Campaña  ---
id_campaña
social_Mexico            317099.52
social_Argentina         306402.46
organic_Colombia         305668.68
organic_Mexico           305173.48
paid_search_Mexico       303502.48
organic_Argentina        302690.85
social_Colombia          294541.23
paid_search_Colombia     281457.87
paid_search_Argentina    278127.86
Name: gasto, dtype: float64


---

## 🔹Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.

⚙️ **Conexión a la base de datos:** Se ha establecido la conexión exitosa con la base de datos de producción para extraer los eventos reales de la plataforma y procesar el flujo de usuarios.

---

**📊 Parte 1: Construcción del funnel**

Para entender el volumen de tráfico, hemos mapeado el recorrido completo del usuario desde su ingreso hasta la conversión final:

* **¿Cuántos usuarios llegan a cada etapa?** 👥
    * **first_visit**: 7,796 usuarios (Nuestra base total / 100%)
    * **select_item**: 7,582 usuarios
    * **add_to_cart**: 7,634 usuarios
    * **begin_checkout**: 7,208 usuarios
    * **add_payment_info**: 6,250 usuarios
    * **purchase**: 6,240 usuarios

* **Orden lógico del flujo**: 🔄 Los eventos se han organizado siguiendo el "happy path" del cliente:  
  `first_visit` ➡️ `select_item` ➡️ `add_to_cart` ➡️ `begin_checkout` ➡️ `add_payment_info` ➡️ `purchase`.

---

**📉 Parte 2: Análisis de conversión**

Analizando los resultados obtenidos, podemos extraer las siguientes conclusiones clave:

* **Tasa de conversión por etapa (respecto al inicio):** 📈
    * El **97.26%** de los visitantes iniciales selecciona al menos un producto.
    * El **92.46%** tiene la intención de pago al iniciar el checkout.
    * El **80.17%** logra completar el registro de su método de pago.

* **¿En qué etapa se pierde la mayor cantidad de usuarios?** ⚠️
    * El punto crítico de fricción ocurre entre **`begin_checkout`** y **`add_payment_info`**. 
    * Se observa una caída significativa de **12.29 puntos porcentuales** en la conversión acumulada. Esto sugiere posibles problemas en la carga del formulario de pago o falta de opciones de pago preferidas por el usuario.

* **¿Cuál es la tasa de conversión final?** 🏆
    * La tasa de conversión total del funnel es del **80.04%**. Es decir, de cada 10 usuarios que inician su visita, aproximadamente 8 logran concretar la compra con éxito.

---

In [18]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [19]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head()

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [20]:
# 1. Agrupamos por nombre_evento y contamos usuarios únicos
query_funnel = events.groupby('nombre_evento')['id_usuario'].nunique().reset_index()
query_funnel.columns = ['nombre_evento', 'usuarios_unicos']

# 2. Definimos el orden lógico
orden_pasos = [
    'first_visit', 
    'select_item',      
    'add_to_cart', 
    'begin_checkout',   
    'add_payment_info', 
    'purchase'
]

# 3. Aplicamos el orden categórico
query_funnel['nombre_evento'] = pd.Categorical(query_funnel['nombre_evento'], categories=orden_pasos, ordered=True)

# 4. Ordenamos y eliminamos cualquier residuo que no esté en nuestra lista
query_funnel = query_funnel.dropna(subset=['nombre_evento']).sort_values('nombre_evento').reset_index(drop=True)

# 5. Cálculo de la Tasa de Conversión (Conversion Rate)
total_inicial = query_funnel['usuarios_unicos'].values[0]
query_funnel['conversion_rate'] = (query_funnel['usuarios_unicos'] / total_inicial) * 100

display(query_funnel)

,nombre_evento,usuarios_unicos,conversion_rate
0,first_visit,7796,100.000000
1,select_item,7582,97.255003
2,add_to_cart,7634,97.922011
3,begin_checkout,7208,92.457671
4,add_payment_info,6250,80.169318
5,purchase,6240,80.041047


In [21]:
# PARTE 1: Totales del funnel
# ======================

query_totals = '''
SELECT 
    nombre_evento, 
    COUNT(DISTINCT id_usuario) as usuarios_unicos
FROM events
GROUP BY nombre_evento
ORDER BY 
    CASE 
        WHEN nombre_evento = 'first_visit' THEN 1
        WHEN nombre_evento = 'select_item' THEN 2
        WHEN nombre_evento = 'add_to_cart' THEN 3
        WHEN nombre_evento = 'begin_checkout' THEN 4
        WHEN nombre_evento = 'add_payment_info' THEN 5
        WHEN nombre_evento = 'purchase' THEN 6
        ELSE 7
    END;
'''

totals = pd.read_sql(query_totals, con=engine)
totals

,nombre_evento,usuarios_unicos
0,first_visit,7796
1,select_item,7582
2,add_to_cart,7634
3,begin_checkout,7208
4,add_payment_info,6250
5,purchase,6240


In [22]:
# PARTE 2: Conversiones
# ======================

query_conversion = '''
WITH base_usuarios AS (
    SELECT 
        nombre_evento, 
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    GROUP BY nombre_evento
),
total_home AS (
    SELECT usuarios_unicos 
    FROM base_usuarios 
    WHERE nombre_evento = 'first_visit'
)
SELECT 
    b.nombre_evento,
    b.usuarios_unicos,
    ROUND((b.usuarios_unicos * 100.0) / (SELECT usuarios_unicos FROM total_home), 2) as conversion_rate
FROM base_usuarios b
ORDER BY 
    CASE 
        WHEN b.nombre_evento = 'first_visit' THEN 1
        WHEN b.nombre_evento = 'select_item' THEN 2
        WHEN b.nombre_evento = 'add_to_cart' THEN 3
        WHEN b.nombre_evento = 'begin_checkout' THEN 4
        WHEN b.nombre_evento = 'add_payment_info' THEN 5
        WHEN b.nombre_evento = 'purchase' THEN 6
        ELSE 7
    END;
'''

conversion = pd.read_sql(query_conversion, con=engine)
conversion

,nombre_evento,usuarios_unicos,conversion_rate
0,first_visit,7796,100.00
1,select_item,7582,97.26
2,add_to_cart,7634,97.92
3,begin_checkout,7208,92.46
4,add_payment_info,6250,80.17
5,purchase,6240,80.04


---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse y medir la fidelidad hacia RappiPlus.

**Tablas utilizadas:**
- `users`: Contiene la fecha de registro original de cada usuario.
- `user_activity`: Contiene el detalle de las interacciones posteriores al registro.

---

### 📑 Metodología de Análisis

1. **Identificación de Cohortes:** Se agrupó a los usuarios según su **mes de registro** (de enero a mayo de 2025) para observar su comportamiento a lo largo del tiempo.
2. **Cálculo de Retención Semanal:** Se midió cuántos usuarios únicos se mantuvieron activos en periodos de 7 días:
   - `retenido_w1`: Actividad entre los días 7 y 13.
   - `retenido_w2`: Actividad entre los días 14 y 20.
   - `retenido_w3`: Actividad entre los días 21 y 27.
3. **Métricas de Salud (Porcentajes):** Se calculó la efectividad de la retención dividiendo los usuarios activos entre el total inicial de la cohorte (`semana_1`, `semana_2`, `semana_3`).

> **Nota técnica:** Se aseguró la integridad de los datos normalizando las fechas con `CAST(fecha_registro AS DATE)` y utilizando `TO_CHAR` para una visualización limpia de los meses.

---

### 📉 Resultados y Conclusiones

Basado en el análisis de las cohortes de 2025, observamos los siguientes hallazgos:

* **Estabilidad del Producto:** ⚖️ La retención en la **Semana 1** se mantiene sólida por encima del **41%** en todas las cohortes, lo que indica una buena primera impresión del servicio.
* **Fidelización a Largo Plazo:** 💎 Sorprendentemente, la retención no presenta caídas drásticas hacia la **Semana 3**. De hecho, la cohorte de **febrero** muestra una retención del **43.98%** en su tercera semana, superando incluso su desempeño de la semana 1.
* **Conclusión General:** 🏆 RappiPlus presenta una retención sumamente saludable. Mantener tasas superiores al **40%** después de 21 días es un indicador claro de que los usuarios encuentran un valor recurrente en la suscripción, lo que reduce el costo de adquisición de clientes a largo plazo.

---

In [23]:
# Explorar tabla users
# ======================
query_users = '''
SELECT 
    id_usuario, 
    CAST(fecha_registro AS DATE) as fecha_registro
FROM users;
'''

users = pd.read_sql(query_users, con=engine)
users.head(3)

,id_usuario,fecha_registro
0,user_0,2025-01-29
1,user_1,2025-01-07
2,user_2,2025-03-12


In [24]:
# Explorar tabla user_activity
# ======================
query_user_activity = '''
SELECT 
    id_usuario, 
    fecha_actividad, 
    dias_despues_registro, 
    activo
FROM user_activity;
'''

user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(3)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1


In [25]:
# Retención por cohortes (Versión con formato de fecha limpio)
# ======================

query_cohort_retention_final = '''
WITH cohort_base AS (
    SELECT 
        id_usuario,
        -- Formateamos la fecha directamente en SQL para que solo muestre YYYY-MM
        TO_CHAR(DATE_TRUNC('month', CAST(fecha_registro AS DATE)), 'YYYY-MM') as mes_cohorte
    FROM users
),
retencion_calculada AS (
    SELECT 
        c.mes_cohorte,
        c.id_usuario,
        MAX(CASE WHEN a.dias_despues_registro BETWEEN 7 AND 13 AND a.activo = 1 THEN 1 ELSE 0 END) as w1,
        MAX(CASE WHEN a.dias_despues_registro BETWEEN 14 AND 20 AND a.activo = 1 THEN 1 ELSE 0 END) as w2,
        MAX(CASE WHEN a.dias_despues_registro BETWEEN 21 AND 27 AND a.activo = 1 THEN 1 ELSE 0 END) as w3
    FROM cohort_base c
    LEFT JOIN user_activity a ON c.id_usuario = a.id_usuario
    GROUP BY c.mes_cohorte, c.id_usuario
)
SELECT 
    mes_cohorte,
    COUNT(DISTINCT id_usuario) as usuarios_iniciales,
    COUNT(DISTINCT CASE WHEN w1 = 1 THEN id_usuario END) as retenido_w1,
    COUNT(DISTINCT CASE WHEN w2 = 1 THEN id_usuario END) as retenido_w2,
    COUNT(DISTINCT CASE WHEN w3 = 1 THEN id_usuario END) as retenido_w3,
    ROUND(COUNT(DISTINCT CASE WHEN w1 = 1 THEN id_usuario END) * 100.0 / COUNT(DISTINCT id_usuario), 2) as semana_1,
    ROUND(COUNT(DISTINCT CASE WHEN w2 = 1 THEN id_usuario END) * 100.0 / COUNT(DISTINCT id_usuario), 2) as semana_2,
    ROUND(COUNT(DISTINCT CASE WHEN w3 = 1 THEN id_usuario END) * 100.0 / COUNT(DISTINCT id_usuario), 2) as semana_3
FROM retencion_calculada
GROUP BY mes_cohorte
ORDER BY mes_cohorte;
'''

cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohorte_final

,mes_cohorte,usuarios_iniciales,retenido_w1,retenido_w2,retenido_w3,semana_1,semana_2,semana_3
0,2025-01,1627,697,668,656,42.84,41.06,40.32
1,2025-02,1444,611,609,635,42.31,42.17,43.98
2,2025-03,1636,677,705,690,41.38,43.09,42.18
3,2025-04,1606,680,697,663,42.34,43.40,41.28
4,2025-05,1687,695,676,706,41.20,40.07,41.85


---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
    - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**
3. **Aplicar el test estadístico adecuado**
4. **Interpretar el resultado** ---
### Hipótesis estadística

- **$H_0$ (Hipótesis nula):** No existe una diferencia significativa entre las tasas de conversión del grupo de control y el de tratamiento.
- **$H_1$ (Hipótesis alternativa):** Existe una diferencia estadísticamente significativa entre las tasas de conversión de ambos grupos.
   
**Test estadístico:** Prueba Z para proporciones (Z-test for proportions).  
**Nivel de significancia alpha:** 0.05 (5%).

---

### Interpretación de resultados

Tras ejecutar el análisis con los datos del experimento, se obtuvieron los siguientes valores clave:

* **Estadístico Z:** 0.8133
* **P-value:** 0.4161

**Conclusión:** Dado que el **p-value (0.4161)** es mayor al nivel de significancia establecido (**0.05**), **no se rechaza la hipótesis nula**. Esto significa que no hay evidencia estadística suficiente para afirmar que el cambio en la UI del checkout haya generado un impacto real en la conversión de los usuarios. La diferencia observada entre el grupo de control (15.69%) y tratamiento (16.29%) se atribuye al azar.

In [26]:
import pandas as pd
from statsmodels.stats.proportion import proportions_ztest

# 1. Cargar el dataset
url = 'https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv'
df_experiment = pd.read_csv(url)

# 2. Exploración rápida por variante
resumen = df_experiment.groupby('variante')['convirtio'].agg(['count', 'sum', 'mean'])
resumen.columns = ['usuarios_totales', 'conversiones', 'tasa_conversion']
display(resumen)

,usuarios_totales,conversiones,tasa_conversion
variante,,,
control,4965,779,0.156898
tratamiento,5035,820,0.162860


In [27]:
# Éxitos (conversiones)
successes = [resumen.loc['tratamiento', 'conversiones'], resumen.loc['control', 'conversiones']]

# Observaciones (totales)
nobs = [resumen.loc['tratamiento', 'usuarios_totales'], resumen.loc['control', 'usuarios_totales']]

print(f"Éxitos (Tratamiento, Control): {successes}")
print(f"Muestra total (Tratamiento, Control): {nobs}")

Éxitos (Tratamiento, Control): [820, 779]
Muestra total (Tratamiento, Control): [5035, 4965]


In [28]:
from statsmodels.stats.proportion import proportions_ztest

# Ejecutar el test
z_stat, p_value = proportions_ztest(successes, nobs)

print(f"Estadístico Z: {z_stat:.4f}")
print(f"P-value: {p_value:.4f}")

# Interpretación lógica
alpha = 0.05
if p_value < alpha:
    print("\n✅ CONCLUSIÓN: La diferencia es ESTADÍSTICAMENTE SIGNIFICATIVA.")
    print("El cambio en la UI impactó positivamente la conversión.")
else:
    print("\n❌ CONCLUSIÓN: La diferencia NO es estadísticamente significativa.")
    print("No hay evidencia suficiente para asegurar que el cambio de UI sea mejor que el actual.")

Estadístico Z: 0.8133
P-value: 0.4161

❌ CONCLUSIÓN: La diferencia NO es estadísticamente significativa.
No hay evidencia suficiente para asegurar que el cambio de UI sea mejor que el actual.


---

## 🔹Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

# 🚀 Entrega de Proyecto Final: Análisis de Rentabilidad RappiPlus

**Analista:** Roberto Carlos Briones Acosta  
**Herramientas:** Power BI, Python (Pandas/Jupyter), SQL  

---

### 📂 Acceso a los Entregables

Para una revisión detallada del proyecto, se proporcionan los siguientes accesos directos:

| Recurso | Descripción | Enlace de Acceso |
| :--- | :--- | :--- |
| **📊 Dashboard Interactivo** | Archivo ejecutable (.pbix) con el modelo de datos y visualizaciones. | [Descargar Dashboard](https://drive.google.com/file/d/15AyKoFI_-KACzRmYWplxJKopFBmJlwUM/view?usp=sharing) |
| **📁 Repositorio de Datos** | Carpeta con datasets limpios (CSVs) y documentación técnica. | [Explorar Carpeta de Proyecto](https://drive.google.com/drive/folders/1ZUQ2IWKc6p2bjCsxItVs18db9R5yj9tG?usp=sharing) |

---

### 📝 Notas de la Entrega:
* **Modelo de Datos:** Se ha configurado una arquitectura en estrella para optimizar el rendimiento del filtrado cruzado por país y categoría.
* **Integridad:** Los datos han sido procesados y validados previamente en Python para asegurar la consistencia de los KPIs de rentabilidad mostrados.